cpu/stm32f1: modernized peripheral GPIO driver

haukepetersen opened the pull request:
One more CPU on the bright side :-)

Code size before (`tests/periph_gpio`, `iot-lab_M3`):
```
   text	   data	    bss	    dec	    hex
  16492	    120	   2888	  19500	   4c2c
```
and after:
```
   text	   data	    bss	    dec	    hex
  15244	    120	   2888	  18252	   474c
```
**Savings ~1,2k**

kaspar030 commented:
nice!

PeterKietzmann commented on cpu/stm32f1/include/periph_cpu.h:
```diff
@@ -19,13 +19,98 @@
+/** @} */
+
+/**
+ * @brief   Available ports on the STM32F4 family
```
stm32f**1** family

PeterKietzmann commented on cpu/stm32f1/include/periph_cpu.h:
```diff
@@ -19,13 +19,98 @@
+/**
+ * @brief   Configure the alternate function for the given pin
+ *
+ * @note    This is meant for internal use in STM32F4 peripheral drivers only
```
stm32f**1** family

PeterKietzmann commented on cpu/stm32f1/periph/gpio.c:
```diff
@@ -30,1015 +30,190 @@
 
-static gpio_state_t config[GPIO_NUMOF];
+/**
+ * @brief   Allocate memory for on callback and argument per EXTI channel
```
This reads strange somehow. Maybe remove "on" or add a minus, e.g.: "on-callback"

PeterKietzmann commented on cpu/stm32f1/periph/gpio.c:
```diff
@@ -30,1015 +30,190 @@
+    /* enable the clock for the selected port */
+    RCC->APB2ENR |= (RCC_APB2ENR_IOPAEN << _port_num(pin));
+    /* clear configuration */
+    port->CR[pin_num >> 3] &= ~(0xf << ((pin_num & 0x7) * 4));
```
Probably I'm just at loss currently. I'm wondering why you shift by `3` and not by `4` here. You want to distinguish between the `CR-low` and the `CR-high` registers (by indexing with `0` or `1`). And AFAIK you want to have `CR[0]` if `pin_num < 16` and  `CR[1]` if `pin_num >=16`. Now just two examples:

```
pin_num = 0d15 = 0b1111
pin_num >> 3 = 0d1 = 0b0001
```

```
pin_num = 0d16 = 0b10000
pin_num >> 3 = 0d2 = 0b0010
```

Where am I mistaken?

PeterKietzmann replied:
IIRC it's the same reason why you do `&0x7` and I think we already had that discussions but I forgot the outcome. Sorry...

haukepetersen replied:
you are almost right. I but the CPU only has 16 pins per port, so I want to distinguish between `x < 8` and `x >= 8`... So look at this:
```
pin_num = 0xd7 = 0b0111
pin_num >> 3 = 0d0 = 0b0000
```
```
pin_num = 0d8 = 0b1000
pin_num >> 3 = 0d1 = 0b001
```

PeterKietzmann replied:
ahja, remember it :-)

PeterKietzmann commented on cpu/stm32f1/periph/gpio.c:
```diff
@@ -30,1015 +30,190 @@
-    return;
+    (void)pin;
+    (void)af;
+    /* TODO: implement and use this in other peripheral drivers */
```
Is this dependent for other peripheral drivers or didn't you want to implement this function right now :-) ?

haukepetersen replied:
The correct comment must be "use this in other peripheral drivers"...

PeterKietzmann commented on cpu/stm32f1/periph/gpio.c:
```diff
@@ -30,1015 +30,190 @@
-    EXTI->IMR &= ~(1 << exti_line);
-
-    return;
+    EXTI->IMR |= (1 << _pin_num(pin));
```
Nice reduction of code lines!

PeterKietzmann commented on cpu/stm32f1/periph/gpio.c:
```diff
@@ -30,1015 +30,190 @@
 }
 
-void gpio_write(gpio_t dev, int value)
+void gpio_write(gpio_t pin, int value)
```
I never really understood why we need this function.. Anyway, it's not related to this PR

haukepetersen replied:
this is very useful when setting a pin value programmatical (is this an English word?!) - see #3538 for example...

PeterKietzmann commented:
The changes look valid to me at a first glance. I don't have all boards here to test, just the iot-lab_M3. Will do now

PeterKietzmann commented:
The transceiver driver seems so work correctly on the iot-lab_M3 node. Does anyone know the pin-setup of the [iot-lab connector](github.com/iot-lab/iot-lab/wiki/Docs/openm3-schematics.pdf)? Can anyone test this on spark-core and fox?

PeterKietzmann commented:
Ok, how do we proceed? @haukepetersen did you test it on spark-core and/or fox boards? We don't have those...

haukepetersen commented:
I will try it out and fix the comments tomorrow.

haukepetersen commented:
addressed comments and added missing implementation of the `init_af` function. Will make a follow up PR to adjust the other peripheral drivers accordingly.

haukepetersen commented:
I am un-able to test the spark-core, my board seems to be electrically broken...

haukepetersen commented:
found one more bug in the `_port_pin()` function, a `-2` offset was missing...

PeterKietzmann commented:
Why is this offset needed? 
I tested this PR again on iot-lab_M3 with *tests/driver_at86rf2xx*  and *tests/periph_gpio*. It seems to be fine.

haukepetersen commented:
Without the offset, Port A would lead to `2` -> which breaks the clock initialization. This just did not show, as the clocks were also activated hard-coded by other peripheral drivers...

haukepetersen commented:
two more fixes: 
- removed the GPIO_xx defines from the iot-labs periph_conf
- enable GPIO clock in `gpio_init_af` function

PeterKietzmann commented on cpu/stm32f1/include/periph_cpu.h:
```diff
@@ -19,13 +19,88 @@
+typedef enum {
+    GPIO_NOPULL = 4,        /**< do not use internal pull resistors */
+    GPIO_PULLUP = 9,        /**< enable internal pull-up resistor */
+    GPIO_PULLDOWN = 8       /**< enable internal pull-down resistor */
```
Could you shortly explain these values? Maybe referring to the [reference manual](http://www.st.com/st-web-ui/static/active/en/resource/technical/document/reference_manual/CD00171190.pdf) (sec.9.2.1, p. 171 or table 20, p.161)? I think it's much faster than figuring this out by myself

haukepetersen replied:
Sure: To understand them, have a look at the gpio_init function and how they are used. I made a bit-map table with the configuration values for the CR[L|H] register and then tried to map the values of DIR and PP to this table. The values you see here is what you get.

PeterKietzmann replied:
Ah, I totally overlooked ` if (dir == GPIO_DIR_OUT)` there

PeterKietzmann commented:
@haukepetersen on iot-lab_M3 I tested tests/periph_gpio, tests/driver_at86rf2xx, examples/default with success. Travis gives us green light. I think no one else will test this on fox and spark-core. So this is an ACK. Please squash your latest commits (or at least remove the "s")

haukepetersen commented:
squashed. Waiting for Travis once more to make sure I did not break anything while squashing...

PeterKietzmann commented:
Let's do this. ACK and go
